### ISSUE TITLE
dart2js vs. DDC: How to cast List<dynamic> Promise callbacks (specific: navigator.usb.getDevices callback)

### ISSUE THREAD

**Issue by trott13** [dart-lang/sdk#39147]
I'm trying to call navigator.usb.getDevices()

This is my Promise type:
`
@JS()
@anonymous
class Promise {
    external void then(Function onFulfilled, Function onRejected);
}
`
and this is the result class for USBDevice:
`
@JS()
@anonymous
abstract class USBDevice {
    // more or less a direct copy of "§6 USBDevice" of
    // https://wicg.github.io/webusb/

    external int get usbVersionMajor;
    external int get usbVersionMinor;
    external int get usbVersionSubminor;
    external int get deviceClass;
    external int get deviceSubclass;
    external int get deviceProtocol;
    external int get vendorId;
    external int get productId;
    external int get deviceVersionMajor;
    external int get deviceVersionMinor;
    external int get deviceVersionSubminor;
    external String get manufacturerName;
    external String get productName;
    external String get serialNumber;
    
    external USBConfiguration get configuration;
    external bool get opened;

    external Promise open();
    external Promise close();
    external Promise selectConfiguration(int configurationValue);
    external Promise claimInterface(int interfaceNumber);
    external Promise releaseInterface(int interfaceNumber);
    external Promise selectAlternateInterface(int interfaceNumber, int alternateSetting);

    external Promise transferIn(int endpointNumber, int length);
    external Promise transferOut(int endpointNumber, ByteBuffer data);
    external Promise clearHalt(String direction, int endpointNumber);
    external Promise reset();

    external factory USBDevice({
        int usbVersionMajor,
        int usbVersionMinor,
        int usbVersionSubminor,
        int deviceClass,
        int deviceSubclass,
        int deviceProtocol,
        int vendorId,
        int productId,
        int deviceVersionMajor,
        int deviceVersionMinor,
        int deviceVersionSubminor,
        String manufacturerName,
        String  productName,
        String serialNumber,
        USBConfiguration configuration,
        bool opened,
        Promise open,
        Promise close,
        Promise selectConfiguration,
        Promise claimInterface,
        Promise releaseInterface,
        Promise selectAlternateInterface,
        Promise transferIn,
        Promise transferOut,
        Promise clearHalt,
        Promise reset
    });
}
`
...and this is how I call it:
`
void displayUSBDevices() {
    print("displayUSBDevices");
    usb_getDevices().then(allowInterop((List<dynamic> devices) { //onFulfilled
        print("usb_getDevices/0: $devices");
        for (final dev in devices) {
            //USBDevice d = dev;
            USBDevice d = dev as USBDevice;
            //USBDevice d = USBDevice(dev);
            //USBDevice d = dev.cast<USBDevice>();
            print("usb_getDevices/2: $d");
            print("productName: ${d.productName}");
        }
    }), allowInterop((error) {  // onRejected
        print("usb_getDevices: $error");
    }));
}
`

This works using the DDC, but not using dart2js.
Then I get something like:

> 
> displayUSBDevices
> 
> js_primitives.dart:47 usb_getDevices/0: [Instance of 'f']
> js_helper.dart:1340 Uncaught (in promise) CastError: Instance of 'f': type 'f' is not a subtype of type 'minified:tR'
>     at Object.d (http://localhost:8000/kiosk/main.dart.js:1702:3)
> ....
> 

So, the question is: How do I cast the resulting `List<USBDevice>` correctly,
even when using dart2js (that is: In production mode)?

Using:
[✓] Flutter (Channel master, v1.10.15-pre.186, on Mac OS X 10.14.6 18G95, locale de-DE)

**Comment by trott13** [comment]
Please note: To make this working in DDC, I need to change the parameter type
from `(List<dynamic> devices)` to `(List<USBDevice> devices)`

**Comment by natebosch** [comment]
For future issues, use 3 ticks for code fences. https://help.github.com/en/github/writing-on-github/creating-and-highlighting-code-blocks

Note that unless you are constructing a `USBDevice` from dart code (which seems unusual) you don't need the factory constructor or the `@anonymous` annotation.

It is _not_ safe to drop `allowInterop`, even when using DDC, and in the future this will become a runtime error there as well.

A `List<dynamic>` can't be cast to a `List<Something>` with `as`. You can use [`list.cast<Something>()`](https://api.dart.dev/stable/2.6.0/dart-core/List/cast.html)  or [`List<Something>.from(otherList)`](https://api.dart.dev/stable/2.6.0/dart-core/List/List.from.html).

The version with a callback taking a `List<dynamic>` should work in both DDC and Dart2js.

**Comment by trott13** [comment]
Thanks for the comments. I changed my code to:
```
void displayUSBDevices() {
    print("displayUSBDevices");
    usb_getDevices().then(allowInterop((List<dynamic> devices) { //onFulfilled
        print("usb_getDevices/0: $devices");
        var devs = List<USBDevice>.from(devices);
        for (final dev in devs) {
            print("usb_getDevices/1: $dev");
            USBDevice d = dev;
            print("usb_getDevices/2: $d");
            print("productName:      ${d.productName}");
        }
    }), allowInterop((error) {  // onRejected
        print("usb_getDevices: $error");
    }));
}
```

This gives me this output & error in DDC:
> displayUSBDevices
> js_primitives.dart:30 usb_getDevices/0: [[object USBDevice]]
> errors.dart:145 Uncaught (in promise) Error: Expected a value of type 'USBDevice', but got one of type '_USBDevice'

And this output & error in dart2js:
> displayUSBDevices
> js_primitives.dart:47 usb_getDevices/0: [Instance of 'f']
> js_primitives.dart:47 usb_getDevices/1: Instance of 'f'
> js_primitives.dart:47 usb_getDevices/2: Instance of 'f'
> main.dart.js:2485 Uncaught (in promise) TypeError: J.ay(...).gAH is not a function

**Comment by natebosch** [comment]
> errors.dart:145 Uncaught (in promise) Error: Expected a value of type 'USBDevice', but got one of type '_USBDevice'

Hmm, this is interesting. I think it might be treating this first as an SDK defined `_USBDevice` which might be preventing you from wrapping it with your own JS interop.

cc @sigmundch - we will need to dig on this. I don't know of a workaround in the short term.

**Comment by sigmundch** [comment]
> I think it might be treating this first as an SDK defined _USBDevice which might be preventing you from wrapping it with your own JS interop.

Indeed, types declared in `dart:html` as a native type cannot be accessed via js-interop at the moment. The only ways we have to allow it have the disadvantage that it can cause many more issues (like name collisions between unrelated js-interop types).

/cc @terrylucas - in this case, it appears the classes are private. If so, should we either expose them or delete them? I'd be inclined to delete them from dart:html, since they are not used today and doing so will enable users to write their js-interop layer on top.

**Comment by trott13** [comment]
Is there something we can do or use as a workaround?
Please note, that this is a showstopper for our project.

**Comment by sigmundch** [comment]
Sorry for the slow response, @terrylucas was out for a few days. He believes we can delete the conflicting definition so we can unblock you. This however wont be available until a future release of the dart sdk because the change is inside an sdk library. Once that's out, your approach with the `@JS` class should start working.

Meanwhile, one alternative I can think of is to use js-interop without the interfaces, but treating the objects more opaquely by accessing their internal fields with helper methods in `js_util`.

For example:
```dart
usb_getDevices().then(allowInterop((List<dynamic> devices) { //onFulfilled
        print("usb_getDevices/0: $devices");
        for (dynamic dev in devs) {
            print("productName: ${getProperty(dev, "productName")}");
```

Here `getProperty` is a helper method in `package:js/js_util.dart`.

**Comment by trott13** [comment]
Yes, this helps a lot! Thanks!

**Comment by sigmundch** [comment]
Sorry this issue hasn't been addressed yet. I recently saw other reports of this issue.

@srujzs - do you know how feasible is to remove the USB* APIs?\

**Comment by srujzs** [comment]
We might be able to just ignore the `modules/webusb` IDL folder entirely in our generation.

Do you know if other reports are on the same API? We have a list of interfaces we suppress by marking them as private classes that we might be able to just not generate at all if they cause similar issues.

**Comment by sigmundch** [comment]
> Do you know if other reports are on the same API? 

I believe so, the error they reported was of this form:
```
TypeError: Instance of 'USBDevice': type 'Interceptor' is not a subtype of type 'UsbDevice'
```

**Comment by refi64** [comment]
Here's a far smaller repo I've run into:

```dart
@JS()
library bug;

import 'dart:html';
import 'package:js/js.dart';

@JS('Headers')
class _JsHeaders {
  external factory _JsHeaders();

  external void append(String name, String value);
}

void main() {
  Headers(); // requires to cause the crash

  var headers = _JsHeaders();
  headers.append('a', 'b');
}
```
